qiskit/utils/deprecation.py

```diff
@@ -275,24 +294,32 @@ def wrapper(*args, **kwargs):
 
 
 def _maybe_warn_and_rename_kwarg(
-    func_name: str,
+    args: tuple[Any, ...],
     kwargs: Dict[str, Any],
     *,
-    old_arg: str,
+    func_name: str,
+    original_func_signature: inspect.Signature,
+    old_arg_name: str,
     new_alias: Optional[str],
     warning_msg: str,
     category: Type[Warning],
     predicate: Optional[Callable[[Any], bool]],
 ) -> None:
-    if old_arg not in kwargs:
+    arg_names_to_values = original_func_signature.bind(*args, **kwargs).arguments
```

Comment:
This can raise a `TypeError` if the user passes in unexpected arguments. The stacktrace might be confusing coming from an internal helper. Maybe we should catch the `TypeError` and enrich it with e.g. mentioning the `func_name`?

Example stacktrace from our test:

```
      File "/Users/arellano/code/qiskit-terra/test/python/utils/test_deprecation.py", line 273, in test_deprecate_arg_runtime_warning
    my_func(bad_arg=0)
      File "/Users/arellano/code/qiskit-terra/qiskit/utils/deprecation.py", line 171, in wrapper
    _maybe_warn_and_rename_kwarg(
      File "/Users/arellano/code/qiskit-terra/qiskit/utils/deprecation.py", line 308, in _maybe_warn_and_rename_kwarg
    arg_names_to_values = original_func_signature.bind(*args, **kwargs).arguments
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/inspect.py", line 3211, in bind
    return self._bind(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/inspect.py", line 3200, in _bind
    raise TypeError(

    TypeError: got an unexpected keyword argument 'bad_arg'
```